optashift-employee-rostering-server/src/main/java/org/optaplanner/openshift/employeerostering/server/solver/WannabeSolverManager.java

```diff
@@ -46,14 +46,27 @@
     private RosterRestService rosterRestService;
 
     private ConcurrentMap<Integer, SolverStatus> tenantIdToSolverStateMap = new ConcurrentHashMap<>();
+    private ConcurrentMap<Integer, Solver<Roster>> tenantIdToSolverMap = new ConcurrentHashMap<>();
 
     @PostConstruct
     public void setUpSolverFactory() {
         solverFactory = SolverFactory.createFromXmlResource(
                 "org/optaplanner/openshift/employeerostering/server/solver/employeeRosteringSolverConfig.xml");
     }
-
-
+    
+    public void terminate(Integer tenantId) {
+        Solver<Roster> solver = tenantIdToSolverMap.get(tenantId);
+        if (null != solver) {
+            tenantIdToSolverMap.remove(tenantId);
+            tenantIdToSolverStateMap.put(tenantId, SolverStatus.TERMINATED);
```

Comment:
(important)
These 2 lines should probably be removed.

This last line is definitely a bug: solver.terminateEarly() is only a signal to stop termination: it returns immediately and when it has returned, the solver might still be running (because that's happening in another thread, the solver thread, while this code executes on a REST thread). So the tenantIdToSolverStateMap shouldn't be adjusted yet. It can still be running and any REST queries for that should still say that it's running.

These 2 lines for the Map changes can be probably removed because the finally inside the executorService.submit block make sure to update those 2 Maps. Normally that should happen withing about 1 millisecond after terminateEarly() was called because OptaPlanner usually reacts that fast on terminateEarly events - if drools score calculation executes submillisecond. In some cases, I 've seen drools score calculation use 200 milliseconds (due to poor DRL) - so it would be a good thing to manually test this by adding a drools rule that does Thread.sleep() or something.

Reply:
Removed those two lines; I modified the drl to do Thread.sleep(); if you terminate it early and hit solve, you get the expected error message "The roster with tenantId (1) is already solving with solverStatus (SOLVING)" (as well as a stack trace), Solve button doesn't change, but timer started (fixed that bug, thanks for suggesting that).